pilot/pkg/xds/proxy_dependencies.go

```diff
@@ -30,26 +30,45 @@ var UnAffectedConfigKinds = map[model.NodeType]sets.Set[kind.Kind]{
 	model.SidecarProxy: sets.New(kind.Gateway, kind.KubernetesGateway),
 }
 
-// ConfigAffectsProxy checks if a pushEv will affect a specified proxy. That means whether the push will be performed
-// towards the proxy.
-func ConfigAffectsProxy(req *model.PushRequest, proxy *model.Proxy) bool {
-	// Empty changes means "all" to get a backward compatibility.
+// filterRelevantUpdates filters PushRequest.ConfigsUpdated so that only configs relevant to the proxy are included,
+// and returns a new PushRequest with the filtered configs.
+func filterRelevantUpdates(proxy *model.Proxy, req *model.PushRequest) *model.PushRequest {
 	if len(req.ConfigsUpdated) == 0 {
-		return true
-	}
-	if proxy.IsWaypointProxy() || proxy.IsZTunnel() {
-		// Optimizations do not apply since scoping uses different mechanism
-		// TODO: implement ambient aware scoping
-		return true
+		return req
 	}
 
+	relevantUpdates := make(sets.Set[model.ConfigKey])
+	changed := false
 	for config := range req.ConfigsUpdated {
 		if proxyDependentOnConfig(proxy, config, req.Push) {
-			return true
+			relevantUpdates.Insert(config)
+		} else {
+			// we have filtered out a config
+			changed = true
 		}
 	}
 
-	return false
+	// If the proxy's service updated, need push for it.
```

Comment:
This is different than the old logic.

Before: if a proxies' ServiceTarget is in ConfigUpdate, we always push
Now: always consider all servicetargets as changed.

TBH on the before, I am not entirely sure why its needed. I guess even if something isn't in its sidecarScope it may be a serviceTarget so we need to incude it? its basically a bug fix for `proxyDependentOnConfig` which probably should have lived in `proxyDependentOnConfig`.

But with the new logic, IIUC, say I have a pod with `service a` selecting it. Now service b changes. I will push service a.

Reply:
Upon further reading: I am wrong, that is not what this is doing and the behavior looks right. It does still seem like maybe proxyDependentOnConfig should have this check, though, but NBD